funsor/minipyro.py

```diff
@@ -345,15 +345,16 @@ def step(self, *args, **kwargs):
 
 # TODO(eb8680) Replace this with funsor.Expectation.
 def Expectation(log_probs, costs, sum_vars, prod_vars):
-    probs = [p.exp() for p in log_probs]
     result = 0
     for cost in costs:
-        result += funsor.sum_product.sum_product(
-                sum_op=funsor.ops.add,
-                prod_op=funsor.ops.mul,
-                factors=probs + [cost],
-                plates=prod_vars,
-                eliminate=prod_vars | sum_vars)
+        log_prob = funsor.sum_product.sum_product(
+            sum_op=funsor.ops.logaddexp,
+            prod_op=funsor.ops.add,
+            factors=log_probs,
+            plates=prod_vars,
+            eliminate=(prod_vars | sum_vars) - frozenset(cost.inputs)
+        )
+        result += funsor.Integrate(log_prob, cost, frozenset(cost.inputs))
```

Comment:
But what if cost is plated? Or batched?

Reply:
Maybe fix the `reduced_vars` arg here and add a final optional `.reduce(ops.add, ..., remaining_plates)`?